aptos-move/block-executor/src/scheduler_status.rs

```diff
@@ -0,0 +1,1387 @@
+    #[test_case(2, 12)]
+    #[test_case(2, 10)]
+    fn stall_pending_scheduling(incarnation: Incarnation, txn_idx: TxnIndex) {
+        let manager = Arc::new(ExecutionQueueManager::new_for_test(10));
+        let status = ExecutionStatus::new_for_test(
+            InnerStatus::new_for_test(StatusEnum::PendingScheduling, incarnation),
+            0,
+            &manager,
+            txn_idx,
+        );
+        // add_stalls work normally, but without changing dependency shortcut flag.
+        manager.add_to_schedule(false, txn_idx);
+        manager.assert_execution_queue(&vec![txn_idx]);
+        assert_ok_eq!(status.add_stall(), true);
+        manager.assert_execution_queue(&vec![]);
+
+        assert_eq!(
+            status.dependency_shortcut.load(Ordering::Relaxed),
+            dependency_flags::DEFER
+        );
+        assert_eq!(status.num_stalls.load(Ordering::Relaxed), 1);
+        assert_ok_eq!(status.add_stall(), false);
+        assert_ok_eq!(status.add_stall(), false);
+        assert_eq!(status.num_stalls.load(Ordering::Relaxed), 3);
+        assert_eq!(
+            status.dependency_shortcut.load(Ordering::Relaxed),
+            dependency_flags::DEFER
+        );
+
+        // remove_stalls work normally, but w.o. changing the dependency shortcut flag.
+        assert_ok_eq!(status.remove_stall(), false);
+        assert_ok_eq!(status.remove_stall(), false);
+        assert_eq!(status.num_stalls.load(Ordering::Relaxed), 1);
+        assert_eq!(
+            status.dependency_shortcut.load(Ordering::Relaxed),
+            dependency_flags::DEFER
+        );
+        manager.assert_execution_queue(&vec![]);
+
+        // Similar (inverted) test flow for remove_stalls.
+        assert_ok_eq!(status.remove_stall(), true);
+        assert_eq!(status.num_stalls.load(Ordering::Relaxed), 0);
+        assert_eq!(
+            status.dependency_shortcut.load(Ordering::Relaxed),
+            dependency_flags::DEFER
+        );
+        manager.assert_execution_queue(&vec![txn_idx]);
+        assert_err!(status.remove_stall());
+    }
+
+    fn set_shortcut_to_safe_or_provided(
+        status: &ExecutionStatus,
+        case: bool,
+        provided_shortcut: u8,
+    ) {
+        status.dependency_shortcut.store(
+            if case {
+                dependency_flags::SAFE
+            } else {
+                provided_shortcut
+            },
+            Ordering::Relaxed,
+        );
+    }
+
+    #[test_case(false)]
+    #[test_case(true)]
+    fn set_executing_flag_err(case: bool) {
+        let status = ExecutionStatus::new(Arc::new(ExecutionQueueManager::new_for_test(10)), 10);
+        // Breaking the invariant, not changing status from PendingScheduling
+        // but updating dependency shortcut flag.
+        set_shortcut_to_safe_or_provided(&status, case, dependency_flags::EXECUTING);
+
+        // Should now panic.
+        assert_err!(status.try_start_executing());
+    }
+
+    #[should_panic]
```

Comment:
Where does the test panic?

Reply:
great catch - it used to assert (and there was even outdated comment that I removed on incarnate). I missed this because I accidentally deleted #[test] annotation instead of should panic 🤦